### ISSUE TITLE
feat(model): load published Flash forward metadata and dense FP8 tensors

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12666]
# feat(model): load published Flash forward metadata and dense FP8 tensors

```routing
lane: model
paths: ["internal/model/config.go", "internal/model/v4_config_admit.go", "internal/model/safetensors_quant.go", "internal/model/v4_dense_quant.go", "internal/model/v4_flash_metadata_test.go", "internal/model/v4_flash_dense_quant_test.go"]
expected_steps: 8
```

## Current state

At public commit `7ac75ceafb8ede8b3a1d432eebcfc87bfad98da2`, the official
Flash expert profile works at component level, but JSON decoding drops
`compress_ratios`, `hc_mult`, `hc_eps`, `hc_sinkhorn_iters`, `o_groups`,
and `o_lora_rank`. The native quant loader expects FP8 scales named
`.weight_scale_inv`; the official checkpoint uses sibling `.scale` tensors
encoded as unsigned E8M0. Its raw attention and shared-expert names do not
enter the quantized matrix store through the existing dense name predicate.

Pinned source:
https://huggingface.co/deepseek-ai/DeepSeek-V4-Flash-0731/tree/7872f01b1d1fe23eabc4c98b48bffcef5a386062
The checked-in official config SHA256 remains
`6c8f3d2d3b48707541b88f32f22ef3f0f8a6b57d8523281e2b8d3cdb0ae9a023`.

## Parent context

Parent: https://github.com/anthony-chaudhary/fak/issues/12635.
Predecessor component: https://github.com/anthony-chaudhary/fak/issues/12659.

## Why this is next

Native attention cannot bind the published checkpoint while its required
configuration is discarded and its dense FP8 scale pairs cannot load.

## Working spine

Official config and safetensors header -> existing native quant loader ->
retained architecture axes and correctly scaled native matrix storage ->
independent numeric regression witness.

## Core through-line

Keep the exact published Flash forward metadata, validate its coupled envelope
before I/O, and consume the official dense FP8/E8M0 pair format using existing
native decoding and Q8 storage. Keep routed FP4 expert tensors and the first
three layers' I64 token-to-expert routing tables lazy through their existing
indexed sources.

The pairing handler applies to Flash's published FP8 roles only. BF16 head,
router and compressor matrices use the existing generic raw-F32 path and have
no fabricated scale companions. This leaf does not preserve dense FP8 storage
or establish full-model logits parity; Q8 is the existing lossy loader format.

## Gold-plating boundary

No full decoder or token-generation claim, mHC implementation, attention
kernel, compression execution, GPU optimization, or V4.1 architecture inference.
The existing Pro expert component behavior stays supported. Do not add a
third-party execution fallback or map V4 weights onto the GQA forward path.

## Value

- Centrality: Core.
- P1 Context: advanced - exact checkpoint metadata survives the native loader.
- P2 Net value: advanced - removes a concrete pre-execution tensor format blocker.
- P3 Adaptation: preserved - unknown/mixed Flash envelopes fail explicitly.
- P4 Operations: advanced - independent bounded fixtures witness scaling and reads.

## Done condition

- [x] [SW-VERIFIED] Official Flash forward metadata survives JSON decoding.
- [x] [SW-VERIFIED] Malformed Flash forward metadata fails before file I/O; Pro controls pass.
- [x] [SW-VERIFIED] The existing native quant loader jointly consumes official-name dense FP8/E8M0 pairs, including scale-first ordering.
- [x] [SW-VERIFIED] Numeric scale/block behavior matches an independent oracle; malformed dtype, published FP8 role shape, missing scales and non-finite values fail explicitly.
- [x] [SW-VERIFIED] FP8 matrices use native Q8 storage under official names; BF16 controls retain their existing representation and routed expert payloads and I64 routing tables remain lazy.
- [x] [SW-VERIFIED] Unretained `mtp.*` payloads are skipped before format decoding; retained unsupported DSpark fails explicitly; existing Pro loader behavior is preserved.
- [x] [SW-VERIFIED] Focused and DeepSeek/V4 regression tests, vet, boundary and leak audits pass.

## Definition of done

The loader returns native matrix data derived from a bounded safetensors fixture
using the exact published naming/scaling contract and preserves the configuration
needed for the next decoder implementation. No generation capability is asserted.

## Witness

An independent test author captures RED on the predecessor, then GREEN on the
implementation. Keep the previously reproduced Qwen35 full-suite failures visible.

## Verifiable Witness

```powershell
$env:FAK_FAST='0'
.\test.ps1 ./internal/model -run 'TestDeepSeekV4Flash(Metadata|Dense)' -count=1 -v
```

Source witness before new tests: `config.go` has no JSON tags for the six fields;
`safetensors_quant.go:414` derives `scaleName := name + "_scale_inv"`, so an
official `.weight`/`.scale` pair cannot satisfy the FP8 loader.

### File:Line seams

- `internal/model/config.go:233-249`: old scalar compression fields.
- `internal/model/v4_config_admit.go:17`: coupled expert admission.
- `internal/model/safetensors_quant.go:31`: `isQuantWeight`.
- `internal/model/safetensors_quant.go:354`: `quantizeNamedTensorsInto`.
- `internal/model/safetensors_quant.go:391`: FP8 companion decoding.

### Blast radius and quarantined fallback

Only public `internal/model` changes. Other model-family loaders retain their
existing scale formats. Unsupported Flash metadata returns an error before I/O;
unimplemented forward stages remain unqualified. No unrelated test is weakened.

## Acceptance gate

Focused metadata/dense tests and broader DeepSeek/V4 tests pass with independent
expected values. Model vet, staged boundary and leak audits pass. Record the full
model suite separately; baseline failures are not evidence of Flash correctness.

## Closure binding

The resolving commit cites this issue and carries `(fak model)`.

## Likely files

The six exact paths in the routing block; one package.

## Lane

Public model; eight steps: fixture/oracle, RED, configuration retention,
admission, paired decoding, native storage, regression verification, landing.

GitHub tracker: https://github.com/anthony-chaudhary/fak/issues/12666

## Verification receipt (2026-09-09)

Resolved and pushed in public commit
[`3532769ab63a8a17dd9c0aa6a7078aad40ed1991`](https://github.com/anthony-chaudhary/fak/commit/3532769ab63a8a17dd9c0aa6a7078aad40ed1991).
Concurrent trunk reconciliation replaced the original guarded landing commit
`b404cafd24bd5c95450585a88ae6c3a68c02c1a7`; independent read-back confirms
byte-identical blobs for all six files and preserves its provenance trailers.
Guarded landing passed prospective and post-merge builds against current trunk;
remote ancestry read-back confirms publication. Context provenance audit is
`SEPARATED`. DOS's subject classifier returned `ABSTAIN` (unrecognized claim
verb); direct commit inspection confirms the six intended implementation/test
paths, and the executed tests provide the scoped correctness witness.

Independent final tests use identical test-file blobs on the predecessor
`7ac75ceafb8ede8b3a1d432eebcfc87bfad98da2` and candidate. The predecessor fails
behaviorally on dropped metadata, official E8M0 scale decoding, role validation,
and eager MTP/I64 reads; the canonical Pro control passes on both. The candidate
passes the focused metadata/dense tests and broader DeepSeek/V4 regression tests
under WSL Go 1.26.6, plus `go vet ./internal/model`.

The full model suite is **RED**, with the same four Qwen35 failures already
reproduced before this work: `TestQwen35PrefillAndStepCPUContinuation`,
`TestQwen35LinearAttnBatchedResumesState`,
`TestQwen35LinearAttnBatchedMatchesScalar`, and
`TestQwen35ChunkedBatchedMatchesScalar`. Existing issues #12433 and #11987 track
those failures. No unrelated test is skipped or weakened.

Staged checks pass for exactly six Go files: all five boundary gates and the
public-copy leak audit. Independent review passes the actual sharded-loader
entry point, including the indexed I64 routing table. The two generated,
unrelated full-suite witness receipts are excluded from this change.

Local logs are under `_scratch/deepseek-v4-flash-12666/`:

| Log | SHA256 |
| --- | --- |
| `red-final-focused.log` | `707dd5a6465ee33d134bddadc96cbe6d4571dded187da3915d77f4b2933d4cbb` |
| `green-focused.log` | `f07aa9c173956f89c4cdfb09dfdf8d8fb4c8a8f3422de6ebade233b616316cdc` |
| `green-deepseek-v4.log` | `28112966397858a801ad73b5aee85f7d812802a39b33ce92b0a2ea84f1b97906` |
| `green-vet.log` | `ba0f13b7256c4d14fad3ac2eccce7dc7b9b95961378459767751281d3746e5b6` |
| `full-package.log` | `1aace949ce8c437be2d0fd1f70db29617d7d0f742019305f1467f215f5ec1cfa` |

This is loader/component evidence only. Full checkpoint loading, native target
token generation, compressed attention, mHC execution, DSpark execution, and
V4.1 compatibility remain unverified.
